pkg/resourcecollector/resourcecollector.go

```diff
@@ -0,0 +1,378 @@
+			continue
+		}
+
+		// Map to prevent collection of duplicate objects
+		resourceMap := make(map[types.UID]bool)
+		for _, resource := range group.APIResources {
+			if !resourceToBeCollected(resource) {
+				continue
+			}
+
+			for _, ns := range namespaces {
+				var dynamicClient dynamic.ResourceInterface
+				if !resource.Namespaced {
+					dynamicClient = r.dynamicInterface.Resource(groupVersion.WithResource(resource.Name))
+				} else {
+					dynamicClient = r.dynamicInterface.Resource(groupVersion.WithResource(resource.Name)).Namespace(ns)
+				}
+
+				var selectors string
+				// PVs don't get the labels from their PVCs, so don't use
+				// the label selector
+				if resource.Kind != "PersistentVolume" {
+					selectors = labels.Set(labelSelectors).String()
+				}
+				objectsList, err := dynamicClient.List(metav1.ListOptions{
+					LabelSelector: selectors,
+				})
+				if err != nil {
+					return nil, err
+				}
+				objects, err := meta.ExtractList(objectsList)
+				if err != nil {
+					return nil, err
+				}
+				for _, o := range objects {
+					runtimeObject, ok := o.(runtime.Unstructured)
+					if !ok {
+						return nil, fmt.Errorf("error casting object: %v", o)
+					}
+
+					collect, err := r.objectToBeCollected(labelSelectors, resourceMap, runtimeObject, ns)
+					if err != nil {
+						return nil, fmt.Errorf("error processing object %v: %v", runtimeObject, err)
+					}
+					if !collect {
+						continue
+					}
+					metadata, err := meta.Accessor(runtimeObject)
+					if err != nil {
+						return nil, err
+					}
+					allObjects = append(allObjects, runtimeObject)
+					resourceMap[metadata.GetUID()] = true
+				}
+			}
+		}
+	}
+
+	err = r.prepareResources(allObjects)
+	if err != nil {
+		return nil, err
+	}
+	return allObjects, nil
+}
+
+// Returns whether an object should be collected or not for the requested
+// namespace
+func (r *ResourceCollector) objectToBeCollected(
+	labelSelectors map[string]string,
+	resourceMap map[types.UID]bool,
+	object runtime.Unstructured,
+	namespace string,
+) (bool, error) {
+	metadata, err := meta.Accessor(object)
+	if err != nil {
+		return false, err
+	}
+
+	// Skip if we've already processed this object
+	if _, ok := resourceMap[metadata.GetUID()]; ok {
+		return false, nil
+	}
+
+	objectType, err := meta.TypeAccessor(object)
+	if err != nil {
+		return false, err
+	}
+
+	switch objectType.GetKind() {
+	case "Service":
+		// Don't migrate the kubernetes service
+		metadata, err := meta.Accessor(object)
+		if err != nil {
+			return false, err
+		}
+		if metadata.GetName() == "kubernetes" {
+			return false, nil
+		}
+	case "PersistentVolumeClaim":
+		metadata, err := meta.Accessor(object)
+		if err != nil {
+			return false, err
+		}
+		pvcName := metadata.GetName()
+		pvc, err := k8s.Instance().GetPersistentVolumeClaim(pvcName, namespace)
+		if err != nil {
+			return false, err
+		}
+		// Only collect Bound PVCs
+		if pvc.Status.Phase != v1.ClaimBound {
```

Comment:
how about a PVC which is stuck in pending because the driver cannot satisfy the request on the source cluster?

There might still be a use case for the PVC spec to get migration so that it can be bound on the destination cluster.

Reply:
There is no way to know whether the PVC is stuck while being bound or if it was just created and will be bound soon. Users might expect different things in both cases. So I'm just being safe and ignoring these for now